Support for llvm 20

kumarak opened the pull request:
The PR changes support to build with llvm 20 and cmake changes to move away from cxx-common dependency

kumarak committed:
add support for building with llvm20

kumarak committed:
Create opaque ptr

kumarak commented:
Moving to draft since I need to fix CI as well.

frabert commented on CMakeLists.txt:
```diff
@@ -50,9 +50,15 @@ find_package(gflags CONFIG REQUIRED)
 llvm_map_components_to_libnames(llvm_libs support core irreader bitreader bitwriter)
-find_package(Clang CONFIG REQUIRED)
+
+find_package(MLIR 20 CONFIG REQUIRED)
```
Does Rellic need MLIR?

ekilmer commented on tools/CMakeLists.txt:
```diff
@@ -57,14 +57,22 @@ add_executable(${RELLIC_XREF}
 )
 
+include(FetchContent)
+FetchContent_Declare(cpp-httplib
+  GIT_REPOSITORY https://github.com/yhirose/cpp-httplib.git
+  GIT_TAG        v0.20.0
+)
+FetchContent_MakeAvailable(cpp-httplib)
+set(CPP_HTTPLIB_INCLUDE_DIRS "${cpp-httplib_SOURCE_DIR}")
+
 find_path(CPP_HTTPLIB_INCLUDE_DIRS "httplib.h")
 target_include_directories(${RELLIC_XREF} PRIVATE ${CPP_HTTPLIB_INCLUDE_DIRS})
 
 target_link_libraries(${RELLIC_XREF}
   PRIVATE
     "${PROJECT_NAME}_cxx_settings"
     "${PROJECT_NAME}"
-    gflags::gflags
+    gflags
 )
```
IMO this should be gated behind an option like
```
option(RELLIC_FIND_ALL_EXTERNAL_DEPENDENCIES "Use find_package for ALL dependencies" OFF)
```
in `cmake/options.cmake`. At least let the user choose to provide their own cpp-httplib, even if it's untested by default. Most users won't care, which is fine, and it'll do the right thing because it's `OFF` by default.

Also, we should use the target `httplib::httplib` instead of `CPP_HTTPLIB_INCLUDE_DIRS`.

ekilmer commented on tools/CMakeLists.txt:
```diff
@@ -20,7 +20,7 @@ target_link_libraries(${RELLIC_DECOMP}
     "${PROJECT_NAME}_cxx_settings"
     "${PROJECT_NAME}"
-    gflags::gflags
+    gflags
```
Why is this changing? gflags provides official support for `gflags::gflags`

artemdinaburg committed:
Add LLVM and CLANG includes to target (#342)

LLVM and CLANG include dirs should be brought in via
target_link_libraries but it would not compile on MacOS for me without
these changes making it explicit.

kumarak committed:
Implementation of pointer inference

kumarak committed:
fix: set flag to use target namespace for gflags
